core-kotlin-modules/core-kotlin-concurrency-4/src/test/kotlin/com/baeldung/threadsafe/ThreadSafeUnitTest.kt

```diff
@@ -0,0 +1,298 @@
+        assertFailsWith<ConcurrentModificationException> {
+            for (item in list) {
+                if (item == 3) {
+                    list.remove(item)
+                }
+            }
+        }
+    }
+
+    @Test
+    fun `test using CopyOnWriteArrayList prevent ConcurrentModificationException`() {
+        val list = CopyOnWriteArrayList(mutableListOf(1, 2, 3, 4, 5))
+
+        for (item in list) {
+            if (item == 3) {
+                list.remove(item)
+            }
+        }
+    }
+
+    @Test
+    fun `test using AtomicInteger to prevent race-condition`() {
+        val list = mutableListOf<Int>()
+        val atomInt = AtomicInteger(0)
+
+        thread {
+            for (i in 1..100) {
+                list.add(i)
+                atomInt.incrementAndGet() // Increment the atomic counter
+            }
+        }.join()
+
+        thread {
+            for (i in 101..200) {
+                list.add(i)
+                atomInt.incrementAndGet()
+            }
+        }.join()
+
+        assertEquals(200, list.size)
+    }
+
+    @Test
+    fun `test using AtomicInteger and synchronized to prevent race-condition reducing potentially deadlock`() {
+        val list = mutableListOf<Int>()
+        val atomInt = AtomicInteger(0)
+
+        val threads = listOf(thread {
+            for (i in 1..100) {
+                synchronized(list) {
+                    list.add(i)
+                    atomInt.incrementAndGet()
+                }
+            }
+        }, thread {
+            for (i in 101..200) {
+                synchronized(list) {
+                    list.add(i)
+                    atomInt.incrementAndGet()
+                }
+            }
+        }, thread {
+            for (i in 201..300) {
+                synchronized(list) {
+                    list.add(i)
+                    atomInt.incrementAndGet()
+                }
+            }
+        })
+
+        threads.forEach { it.join() }
+
+        val actualSize = list.size
+        val countedSize = atomInt.get()
+        logger.info("Actual list size: $actualSize, Counter value: $countedSize")
+
+        if (actualSize != 300 || countedSize != 300) {
+            logger.warn("Possible race condition detected! List size or counter is incorrect.")
+        }
+    }
+
+    @Test
+    fun `test using Collections-synchronizedMap to prevent thread-safety issue`() {
+        val map = Collections.synchronizedMap(HashMap<Int, String>())
+
+        thread {
+            for (i in 1..100) {
+                map[i] = "Thread 1 - $i"
+                Thread.sleep(100) // simulate delay
+            }
+        }.join()
+
+        thread {
+            for (i in 101..200) {
+                map[i] = "Thread 2 - $i"
+            }
+        }.join()
+
+        thread { map.remove(200) }.join()
+
+        assertEquals(199, map.size)
+    }
+
+    @Test
+    fun `test using ConcurrentHashMap for alternative to Collections-synchronizedMap`() {
```

Comment:
This one doesn't make sense, if we are joining each time then it's basically single-threaded + it should be possible to modify concurrently with a ConcurrentHashMap without getting a ConcurrentModificationException